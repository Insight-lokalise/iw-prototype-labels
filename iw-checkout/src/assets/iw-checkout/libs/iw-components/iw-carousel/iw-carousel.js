import React from 'react'
import Slider from 'react-slick' // Slider may require special css if it is within a flex container
import cn from 'classnames'

export function IWCarousel(props) {
    let slides = props.children

    let slidesToShow = (props.settings && props.settings.slidesToShow) || defaultSliderSettings.slidesToShow
    let needMoreSlides = !slides.length || (slides.length && slides.length < slidesToShow)

    let settings = Object.assign({}, defaultSliderSettings, {
        prevArrow: <CustomArrow direction='left' hideArrows={needMoreSlides} />,
        nextArrow: <CustomArrow direction='right' hideArrows={needMoreSlides} />,
        infinite: !needMoreSlides,
    }, props.settings)

    // Let's keep this logic for future reference. Filler slides logic may be helpful
    // if (slides.length && slides.length < slidesToShow) {
    //     const fillerSlides = Array(slidesToShow - slides.length)
    //         .fill('Empty Slide')
    //         .map((v, idx) => <div key={`${v}${idx}`} className='invisible'></div>)
    //     slides = slides.concat(fillerSlides)
    // }

    return (<Slider className={`Carousel ${props.className}`} {...settings} >
        { slides.length
            ? slides
            : <div className='loading'></div> }
    </Slider>)
}

const defaultSliderSettings = {
    slidesToShow: 5,
    infinite: false,
    responsive: [
        {
            breakpoint: 768,
            settings: {
                slidesToShow: 2,
            },
        },
        {
            breakpoint: 1180,
            settings: {
                slidesToShow: 3,
            },
        },
        {
            breakpoint: 1600,
            settings: {
                slidesToShow: 4,
            },
        },
    ],
}

/**
 * To add custom classes to the arrows, we must wrap an element in a custom tag.
 *
 * By destructuring @param currentSlide and @param slideCount from the arguments, we prevent them
 * from filtering as attribtues to the native span tag.
 */
function CustomArrow({ currentSlide, slideCount, direction, hideArrows, ...props }) { // eslint-disable-line no-unused-vars
    return <span {...props}>
        <span className={`${cn({ hide: hideArrows })} ion-ios-arrow-${direction} Carousel__arrow`}></span>
    </span>
}
