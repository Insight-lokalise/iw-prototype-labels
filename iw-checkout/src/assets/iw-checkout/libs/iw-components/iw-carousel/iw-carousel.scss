.slick-prev,
.slick-next {
    position: absolute;
    top: 50%;
    width: 15px;
    height: 40px;
    transform: translate(0, -50%);
    cursor: pointer;
    z-index: 10;
}

.slick-prev {
    left: -15px;
}

.slick-next {
    right: -15px;
    text-align: right;
}

.slick-disabled {
    display: none !important;
}

.Carousel {
    min-height: 270px; // Set height before content is loaded.

    &__arrow {
        font-size: rem-calc(36);
        color: $brand-mediumgray;

        &:hover,
        &:focus {
            color: $brand-crimson;
        }
    }
}
