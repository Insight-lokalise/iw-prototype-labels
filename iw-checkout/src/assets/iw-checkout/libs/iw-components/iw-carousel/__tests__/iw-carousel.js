import React from 'react'
import { shallow, mount } from 'enzyme'

import { IWCarousel } from './../iw-carousel'

/**
 * Tests on this component should test the internal logic for hiding and showing
 * arrows based on slide length and using default/custom arrows.
 *
 * Snapshot testing would be more valuable on the business components that
 * implement this iw-component.
 */
describe('<IWCarousel>', () => {
    it('should hide arrows when slides < slidesToShow', () => {
        const slides = [
            { name: 'Projector', price: 200 },
            { name: 'Screen', price: 60 },
            { name: 'Cable', price: 20 },
        ].map(item => <div className="Test__slide" key={item.price}>
            {item.name} <strong>{item.price}</strong>
        </div>)
        const component = mount(
            <IWCarousel>
                { slides }
            </IWCarousel>
        )
        expect(component.find('.Carousel__arrow.hide').length).toBe(2)
    })

    it('should not hide arrows when slides > slidesToShow', () => {
        const slides = [
            { name: 'Projector', price: 200 },
            { name: 'Screen', price: 60 },
            { name: 'Cable', price: 20 },
        ].map(item => <div className="Test__slide" key={item.price}>
            {item.name} <strong>{item.price}</strong>
        </div>)
        const component = mount(
            <IWCarousel settings={{ slidesToShow: 2 }}>
                { slides }
            </IWCarousel>
        )
        expect(component.find('.Carousel__arrow').length).toBe(2)
        expect(component.find('.Carousel__arrow.hide').length).toBe(0)
    })

    it('should match snapshot', () => {
        const slides = [
            { name: 'Projector', price: 200 },
            { name: 'Screen', price: 60 },
            { name: 'Cable', price: 20 },
        ].map(item => <div key={item.price}>{item.name} <strong>{item.price}</strong></div>)
        const component = shallow(
            <IWCarousel className="Test__carousel">
                { slides }
            </IWCarousel>
        )
        expect(component).toMatchSnapshot()
    })
})
