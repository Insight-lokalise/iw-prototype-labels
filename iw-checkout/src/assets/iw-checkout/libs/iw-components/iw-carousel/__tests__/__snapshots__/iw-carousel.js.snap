// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IWCarousel> should match snapshot 1`] = `
<Slider
  className="Carousel Test__carousel"
  infinite={false}
  nextArrow={
    <CustomArrow
      direction="right"
      hideArrows={true}
    />
  }
  prevArrow={
    <CustomArrow
      direction="left"
      hideArrows={true}
    />
  }
  responsive={
    Array [
      Object {
        "breakpoint": 768,
        "settings": Object {
          "slidesToShow": 2,
        },
      },
      Object {
        "breakpoint": 1180,
        "settings": Object {
          "slidesToShow": 3,
        },
      },
      Object {
        "breakpoint": 1600,
        "settings": Object {
          "slidesToShow": 4,
        },
      },
    ]
  }
  slidesToShow={5}
>
  <div>
    Projector
     
    <strong>
      200
    </strong>
  </div>
  <div>
    Screen
     
    <strong>
      60
    </strong>
  </div>
  <div>
    Cable
     
    <strong>
      20
    </strong>
  </div>
</Slider>
`;
